compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs

```diff
@@ -685,31 +782,64 @@ impl Context {
             }
         }
 
-        let read = self.acir_context.read_from_memory(block_id, &var_index)?;
-        let typ = match dfg.type_of_value(array) {
-            Type::Array(typ, _) => {
-                if typ.len() != 1 {
-                    // TODO(#2461)
-                    unimplemented!(
-                        "Non-const array indices is not implemented for non-homogenous array"
-                    );
+        let mut var_index = if matches!(array_typ, Type::Array(_, _)) {
+            let array_len = dfg.try_get_array_length(array_id).expect("ICE: expected an array");
+            self.get_flattened_index(&array_typ, array_id, array_len, var_index)?
+        } else {
+            var_index
+        };
+
+        let results = dfg.instruction_results(instruction);
+        let res_typ = dfg.type_of_value(results[0]);
+
+        let mut first_elem = true;
+        let value = self.array_get_value(&res_typ, block_id, &mut var_index, &mut first_elem)?;
+
+        self.define_result(dfg, instruction, value);
+
+        Ok(())
+    }
+
+    fn array_get_value(
+        &mut self,
+        ssa_type: &Type,
+        block_id: BlockId,
+        var_index: &mut AcirVar,
+        first_elem: &mut bool,
+    ) -> Result<AcirValue, RuntimeError> {
+        let one = self.acir_context.add_constant(FieldElement::one());
+        match ssa_type.clone() {
+            Type::Numeric(numeric_type) => {
+                if !*first_elem {
```

Comment:
Could you explain what the first element is?
Is it because you store the element sizes at the beginning when you initialise the array?

Reply:
Why do you add one for the other indexes?

Reply:
If we are not on the first element that means we have recursively called `array_get_value` and need to read the other ACIR vars from a nested array as they are represented with a flat structure in our memory blocks. As per the example in the main PR description, if we wish to fetch `x[3].bar.inner` the starting index is 25, but we need to read elements 25, 26, and 27.

Reply:
I get it, thank you for the explanation!
Why don't you rather read on the provided index and then always add one to it at the end?
That way, you do not need first_elem.

Reply:
You are correct! I switched it and it is much cleaner now